pulp_docker/app/tasks/stages/wait_complete.py

```diff
@@ -0,0 +1,38 @@
+import asyncio
+
+from stages import Stage
+
+
+class WaitUntilComplete(Stage):
+    """
+    Does not output until previous stages are complete.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        pending_q = asyncio.Queue()
+        # Wait
+        while True:
+            next_in = await in_q.get()
+            if next_in is None:
+                break
+            await pending_q.put(next_in)
```

Comment:
This will result in all of the declarative objects in the pipeline to be in memory.  Correct?

Reply:
Yes, unfortunately. I am still hoping that the many-to-many saves can rely on the Queue order and won't be a problem. If I'm wring though, the backup plan is to use this stage with a StageGroup and a ConcurrentRunner.

    handle_content = StageGroup([ArtifactDownloader(), SerialSaveArtifact(), ProcessContent(), SerialSave()]
    nested_handle_content = ConcurrentRunner(
        StageGroup([handle_content, handle_content, handle_content, wait, RelateContent])
    )
    pipeline = [TagListStage(), nested_handle_content, ...])

This pattern would force keeping all the declarative objects *for a single Tag* in memory. Of course, it is running concurrently, so it will actually be many tags at once, but the point is that as soon as a Tag and all nested content related to that tag pass to the wait stage, all of it is passed through the remaining pipeline and purged from memory relatively quickly.